Proposal for QuComput framework

Roger-luo opened the pull request:
This a rough draft for my proposal with some more detailed descriptions.

Hope people will give me advice :-)

Commit:
add proposal draft

add init draft of GSoC2016

add init draft of GSoC2016

add init draft of GSoC2016

add init draft of GSoC2016

change file name

add descriptions

add descriptions

add cooling

add pics

correct a word

correct a word

add stabilizer

add stabilizer

correct mistake

change timeline

change timeline

change details

change code style

change code style

change code style

correct an error

correct an error

change docs

add reference

add reference

correct a word

add features

add differences to Language based simulators

changed some prototypes

update

add Luxor.jl

change explaination of goals

add link

correct Kebab

change back to AbstractGates

del paragraph

update

update

add QFT in adiabatic

update AbstractComputModels

reworked type implementation

detialed timeline

timeline

timeline

correct a mistake

del chinese notes

add qpp

add README

add title

add diff

change name

delete a paragraph

amitjamadagni commented:
With reference to the discussion I had with Roger, here is my understanding of what the proposal aims at presenting : 
The broad outline is given a quantum algorithm the package aims at providing various computational models to realize the quantum algorithm with features such as circuit visualization, circuit timing and tweaking the parameters of the model to get deeper insights on working of the algorithm. 
For example, let us assume the quantum fourier transformation (QFT) : this can be realized through a computational model like the adiabatic computational model or the one way computation model. The package aims at providing an interface to tweak the adiabatic parameters if adiabatic computational model is used, or topological parameters if topological model is used, to gain additional insight into the working of a algorithm using a particular model. In addition it aims at providing the end user to generate the circuit diagram, timing analysis of the algorithm when different models are used and also additional features like the noise thresholds of a particular circuit. 

Assuming the above scenario, it would be nice to hear from Roger on the following issues : 
1.  Some comments on the applicability and the use case of such a package ?!
2.  There is a mention of quantum error correction, as from the discussion we had, it would be nice if we could point the algorithms that the package aims to deliver with respect to QECC.
3. Any references to all the computational models that the package aims to deliver would be nice and also it would be great to hear on the timeline that you plan to implement these features so that we dont have a very tight timeline. 

@Roger-luo Please do let me know if I have misunderstood any parts of the proposal !

Roger-luo commented:
@amitjamadagni @yongjianhan

To Reference for Quantum Computational Models:

[1] Adiabatic Quantum Computation[Quantum Computation by Adiabatic Evolution](http://arxiv.org/abs/quant-ph/0001106v1)
[2] [One way and graph state](http://arxiv.org/abs/quant-ph/0602096v1)
[3] [Quantum Circuits](http://www.goodreads.com/book/show/153910.Quantum_Computation_and_Quantum_Information) (it's a book)

To feasibility:
The prototype [Kebab]() have already realized circuits and adiabatic computation, so actually my work on this two features will only be putting these codes in the project's structure and polish them. A rough evaluation for the one-way computation is about 20 days.

The visualization is only about circuits and graph states, which could be simple.(for graph states the GraphPlot.jl have already do the job, so I will use it as dependency)

To time line:
Due to my lack of experience, I may not have a clear thought about time line. I'll update it in proposal after further consideration and discussion with Prof. Han.

To the error correction:
The error correction method is also in the article about one way model I put as reference and Neilson's book with chapter on quantum encoding/error correction(the CSS code chapter).

I'm also considering to reconstruct this python program [QuaEC](https://github.com/cgranade/python-quaec), but I do not have confidence to finish this with other features in 2 months. :-( 

Comments on applicability:
First, the features for circuits visualization is obviously useful for people who need draw pictures for their article or other materials. 
Second, the simulator for different computational models will offer an easier way to simulations for algorithms and experiments. One example is the cooling-assist adiabatic computation referred in Kebab's README.md, this simulation will need both circuits and adiabatic computational model. If there was a framework for this, the work will be much easier.

In addition:
I did not really realize an algorithm on different computational model before, however, they are theoretically proved to be equivalent(they are all universal computational models), and I saw Shor algorithm for adiabatic model and quantum circuits by search in Google. [Quantum Computation and Shor's Factoring Algorithm](https://www.researchgate.net/publication/258098278_Quantum_Computation_and_Shor's_Factoring_Algorithm)

The reason to provide different models is not only for users to play same algorithm on different models, this will provide more flexibility for users, they can test their new theory on the computation models they are going to realize in experiment, or they can even mixed up models as I did with the cooling-assist adiabatic computation, and maybe other applications that I cannot tell at present. 

Simulations can be more useful for experiments and realizations, that's the reason why I mentioned that the approach to the calculation should be as close to experimental equipment as possible. For example, trotter expansion should be used, for we cannot actually add up Hamiltonians in experiments.(at least in our lab's experiments, I do not know if there is other way to add up Hamiltonians in experiments)

Should I add this into proposal markdown file? Please correct me if there is any mistake and if there's something misleading please point it out.

amitjamadagni commented:
> To Reference for Quantum Computational Models:
> 
> [1] Adiabatic Quantum Computation[Quantum Computation by Adiabatic Evolution]> (http://arxiv.org/abs/quant-ph/0001106v1)
> [2] [One way and graph state](http://arxiv.org/abs/quant-ph/0602096v1)
> [3] [Quantum Circuits](http://www.goodreads.com/book/show/153910.Quantum_Computation_and_Quantum_Information) > (it's a book)

The way I understood is that the algorithm and the circuit are universal and we have different computational models to realize the algorithm. If this is the case, what does it imply to assume that quantum circuit is a computational model like the others in this category that is the adiabatic, one way and topological. Also, do you plan to work on topological models ?!

> To time line:
> Due to my lack of experience, I may not have a clear thought about time line. I'll update it in proposal > after further consideration and discussion with Prof. Han.
> I'm also considering to reconstruct this python program [QuaEC](https://github.com/cgranade/python-quaec), but I do not have confidence to finish this with other 
> features in 2 months. :-( 
> 
> In addition:
> I did not really realize an algorithm on different computational model before, however, they are 
> theoretically proved to be equivalent(they are all universal computational models), and I saw Shor 
> algorithm for adiabatic model and quantum circuits by search in Google. [Quantum Computation and > Shor's Factoring Algorithm](https://www.researchgate.net/publication/258098278_Quantum_Computation_and_Shor's_Factoring_Algorithm)
> The reason to provide different models is not only for users to play same algorithm on different 
> models, this will provide more flexibility for users, they can test their new theory on the computation 
> models they are going to realize in experiment, or they can even mixed up models as I did with the 
> cooling-assist adiabatic computation, and maybe other applications that I cannot tell at present. 

Also in our discussions, there was a mention about providing an interface for the user to tweak the various parameters of the computational model. It would be really helpful, if you could detail the parameters that can we tweaked around in different models and how there could be a general design to access the tweaking parameters.

For example, consider the Adiabatic Computational model, what could be the possible parameters that the user gets to tweak to get a better understanding of the algorithm. Also, are there any parameters which are common to all the computational models, if not what could be the design constraints in having a general design. 

Also there was a mention on the timing analysis (do we have something equivalent on the lines of functionality of Quipper, as far as I can understand the functionality of Quipper they support circuit generation and then feed various inputs to the circuits and then compute the time taken for the algorithm to be executed, I might be completely wrong but the ideas run on similar lines). We also had the mention of noise, it would be nice to get an understanding of how noise is realized, I guess it is more a model related property rather than a particular algorithm, or are we dealing with the gate noise
 ?!

It would be really helpful for the community to provide their thoughts on the proposal ! 

@acroy @i2000s @jrevels

acroy commented:
@Roger-luo : First of all, thanks for the proposal and sorry for the late reply. There are a few minor wrinkles, which we can iron out later. I can't say much about the scope and feasibility, but it is great that you have prototypes which show that it basically works.

I think it would be good (for the non-expert reader) to have a paragraph explaining the general idea of quantum simulation and how it fits into JuliaQuantum. As far as I can tell there are the following main ingredients: 1) you need to represent the input/output states, 2) from a graph of gates you might need to construct the Hamiltonian, 3) the input state needs to be propagated through the gates and 4) one needs visualization of the states (and gates). Some questions that come to mind:

1) How are you going to represent the input/output states? Can one profit from the ideas in QuBase?

2) The construction of the Hamiltonian is probably already included in your repo?

3) Will you only use the adiabatic propagator? If you allow for noise/decoherence you will need some other methods as well?

4) How exactly are the states going to be visualized? Can this be used in a more general context?

Overall, I think this would be a cool project to have. There is a large community interested in quantum simulations and maybe we can attract them to use Julia(Quantum).

i2000s commented:
I have been enjoyable reading the proposal and ideas discussed above. Overall, I think this is a project that JuliaQuantum wants to have in reaching out to the fast developing field of quantum information processing. I have the following suggestions, which hopefully can help simplify the planned works a little bit and make this proposal more attractive and generalized. There are some places not clear to me and requiring @Roger-luo to clarify as well. 
1. I feel there are a lot of discussions between @Roger-luo and @amitjamadagni on what this project is about, whether it is going to be something like Quipper or else. I think the concern here is about the difference between **_quantum circuit simulation**_ and **_quantum circuit compiling**_. Long story short, a typical workflow for quantum simulation using quantum circuit is the following: one defines the quantum circuit, implements with quantum gates and other elements, gives an initial state and then simulates the final output being measured. It might be useful to make a circuit CAD tool so that one can easily drag and draw a quantum circuit and then give an input to simulate the output. On the other hand, a typical process of quantum compiling is the following: one defines a Hamiltonian, defines the set of available quantum gates and measurements and then finds the exact or approximate quantum circuit to implement the Hamiltonian. Quipper is more like a pure quantum circuit compiler. Its output is a quantum circuit diagram, and that requires the program to be able to draw the circuit based on the compiling. One can use a compiler to find a corresponding quantum circuit of a target Hamiltonian to implement in experiments. A quantum simulator can also be a combination of a typical quantum compiler and a straightforward quantum circuit simulator, which is like Liquid. Liquid can compile a Hamiltonian into classically simulatable quantum circuits and then run the circuit simulation on computers for the output result. @Roger-luo you may want to specifically clarify what type of simulation functionality and what kind of quantum circuit tool do you want to implement. Without distinguishing if you goal is more like a quantum compiler, a simple simulator based on circuit CAD or a simulator with circuit compiler capability, it could be confusing to read your goal of `make it as close as experiments`. I suggest you to implement the framework of both. I will explain it in the next part.
2. What should be the focus of this project? My suggestion is to focus on implementing a framework of quantum simulations on classical computers using Julia. **_First of all**_, this direction hasn't been implemented that much in the existing JuliaQuantum packages. We need to try and just do it to see what kind of type system works for this direction and what is most important in terms of defining a standard and foundation for further development. The program abstraction of the Julia language, especially its typing system, and its high performance are the biggest advantage (at least to me) to make programming in Julia for simulating quantum processes easier than object-oriented language written packages like QuTIP and faster than functional programming language written packages like Quipper. You may want to highlight this in your proposal. **_Second**_, you may want to cut contents that can go into the base projects off from your proposal. Otherwise, it will loose the focus and seem hard to implement in a short period. **_Third**_, the proposal should include a quantum compiler, circuit/adiabatic/one-way quantum simulators as you preferred, some capability to simulate errors for quantum error correction simulations (illustrate more next), and enough examples and documents. You should be able to make up a schedule for achieving these simplified and focused targets. **_Forth**_, the adiabatic simulations and other simulations on the dynamic processes would be better or more rely on the dynamics solvers we have in Julia. Since the QuDynamics.jl project is still under construction, why not coordinate with @amitjamadagni and let that project be enhanced while you just focus on setting up the framework for quantum computing in Julia? That would be good both for you and the org. 
3. With the suggested focus, I think the type system can be simplified and generalized a little bit. 
   
   i) On the top hierarchy, you need the AbstractCircuit and QuCircuit or Circuit whichever name you prefer. A circuit type should also include the properties like width, depth, number of gates besides what you have defined. The width defines the max number of qubits that a circuit can process, so that you may not need to define the Qubit type for this purpose. The depth defines the number of layer of gates. With those extra properties defined, one can use this program to find the optimal configuration with constant or limited width/depth/number of gates. This is a hot area of study for design optimizations. Also, a circuit should also be able to include **_Qudits**_ rather than just **_Qubits**_ input elements. It has been confirmed that with dimension d>2, a qudit system requires higher threshold for fault-tolerant quantum computing and systems based on qudits are more accessible than qubits for some physics systems. That has been ignored by most previous quantum circuit packages, and should not be hard to be implemented in Julia. For the methods, you should also include taking/feeding/initializing an input state for the circuit, making a gate in error (more see the following discussion on gates), doing a specific measurement and so on. In other word, a circuit should include methods to read in, propagate gates, generate errors and perform error corrections, diagnose in the process, and read out. The missing measurement part is very important. With the measurement method well-defined, this framework may enable quantum process/state tomography, feedback/feedforward control and learning algorithms studies for the future development.
   
   ii) The next level should be AbstractQuGate or AbstractGate whichever you prefer and the specific Gate or QuGate types. It has been defined in the base projects. You can implement some basic gates like CNOT, X, Y, Z, T, S, Toffoli and so on. I want to highlight a few points here. **_First**_, you may want to implement probabilistic error gates. For example, a gate has probability of _p_ to be X-gate and _(1-p)_ as identity. These error gates can be put anywhere on the quantum wires if not correlated or across wires if correlated. These can be used to simulate dephasing, depolarizing, decoherence, loss error and other error channels (see Nielson and Chuang's Quantum Information and Quantum Computation textbook). To my knowledge, this is a very common trick to simulate error threshold and the efficiency of error correction algorithms. @ciaranra may have more comments to guide you on implementing quantum error correction codes and simulations. **_Second**_, based on the famous Gottesman–Knill theorem, stabilizer circuits which relate to Clifford gates can be efficiently simulated on a classical computer, and are the key element for some quantum compiler algorithms. There are tons of algorithms to expand a Hamiltonian and approximate it to Clifford gates so that a classical computer can simulate quantum algorithms. For instance, the recent papers like [this one](http://arxiv.org/abs/1601.07195) and [its followup](http://arxiv.org/abs/1601.07601) may not show you how their software work but these papers and their references can give you some idea on simulating quantum chemistry or quantum processes on a computer cluster with a quantum compiler or with quantum circuits. The top record of simulatable max qubit number is about 49 on the current best supercomputers according to those papers. Liquid package is also compared in these papers as well. So, implementing efficient Clifford gates is a key for a good quantum compiler. Fortunately, one efficient [Clifford gate algorithm](https://github.com/BBN-Q/Cliffords.jl) has been implemented in Julia by our member @blakejohnson and fellows. You may want to include it. Microsoft and others are putting a lot of efforts to optimize their quantum compilers as it is still a big barrier for efficiently simulating quantum algorithms on classical computers. You may want to have a discuss with Prof. Han on whether to implement some compiler algorithms in Julia. But I suppose you want to put some efforts on this direction, like checking [this paper](http://arxiv.org/abs/1411.4028) on approximating a Hamiltonian into a circuit. **_Third**_, it would be nice to figure out a way of programming the gate type to allow a gate element to be easily replaced, added and removed in the process of simulation. This can be useful for simulating gate errors and error correction operators, and for circuit CAD as well. 
   
   iii) I guess other unrelated types like AbstractVector can be replaced by types already defined in the base repos. You may also want to coordinate with @amitjamadagni on implementing Stochastic calculus in Julia within the QuDynamics project. It could be useful for simulating measurements and some errors. Our member @jarthurgross has implemented [one Stochastic master equation solver](https://github.com/CQuIC/pysme) in Python. 
4. On the circuit visualization, you specifically need to clarify how to implement that in Julia clearly. As I have said earlier, you can make a circuit CAD for quantum circuit simulation, which I am not sure how to do it efficiently; or, you can generate a circuit based on quantum compiler algorithms. For the latter one, Liquid is using some LaTeX macros and TikZ to draw quantum circuits and then export the circuits as svg, pdf, eps or any other formats. On the same line, CQuIC's former member Bryan Eastin, Steve T Flammia and current member Travis L Scholten (@Travis-S) have made the well-know [Qcircuit](https://github.com/CQuIC/qcircuit) LaTeX package to draw quantum circuits based on XY-pic. There is also another quantum circuit drawing tool in LaTeX based on TikZ under development in a private repo of [CQuIC](https://github.com/CQuIC). If you are interested in using them, let me know. I can help you get familiar with them and access to the private repo. You may only need to implement a parser to draw the circuits in LaTeX language. I think you have some experience to parse Mathematica expressions to LaTeX. This shouldn't be too hard for you if you want to work on this direction. 
5. Before we further polish the proposal, you may want to correct some small language problems in your proposal. For example, periods, commas and other signs should end with a space before the next phrase starts; a sentence should be ended with a period or other signs not just a return of line or blank; there should be a space before a bracket in a normal sentence; the first letter of a People's name should be capitalized. "Meet up" and "it self" should all be one word; "open source" may be better to be hyphened if being used as an adjective. Also, the name of your proposal file shouldn't have the special character `:` which is not compatible in Windows OS. 

I have commented on the [J/GSoC discussion](https://github.com/JuliaQuantum/JuliaQuantum.github.io/issues/32) thread on helping you coordinate other things. You should contact @acroy or other JuliaQuantum members directly to discuss the possibility of being your mentor for the project. I will also be available if you and Prof. Han want to talk about the project on Skype.

Good luck!
Qi

Roger-luo commented:
I update the Kebab's interface, which may be much better than before. And I changed some of the design in proposal as I found it could be better while reworking the circuit part of Kebab. The newer interface is in the README of Kebab.

@acroy 

#### Input/output states

states are currently an array. I use AbstractVector in Kebab. As it is actually an array, it will be good to just use QuArray instead, if it's better to put it into QuBase.jl. Furthermore a type for Hamiltonians maybe needed in QuBase or other libs, I saw this in QuDynamics, I think it could be nice, if we make it as a public type in QuBase, so I can focus on other things. And it seems there isn't any other requirements at present.

#### construction for Hamiltonian

this could be a misleading part that I didn't explain well. The Hamiltonian constructor is only for the Adiabatic Model, which needs to construct a Hamiltonian, however, this may not be a good way to construct a Hamiltonian in numerical programs, just wrote that to give myself a better interface for writing simulations for Adiabatic models(which comes from the AdiaRoll.jl that I wrote to simulate a work from our lab)

Therefore this Hamiltonian constructor is not a universal tool for other models.

#### Adiabatic propagator

yes, it's necessary to use other methods. But currently, as I'm only aiming to realize the basic features, I haven't considered deeper about this. And for this part I'm planing to use QuDynamics.jl as dependency.

#### visualization

visualization is about two parts:
- for circuits: the visualization is to draw the circuits in the similar style as pics about QFT circuit in proposal this could be a useful part for circuit CAD, but may not be a general visualization method in quantum mechanics.
- for graph states: the visualization is actually to draw the structure of the graph, which have already realized by GraphPlot.jl and what I need to do could only be tagging qubits on the graph. And this state visualization will only make sense for graph states, cause only graph state have a structure of graph.
- Problem on visualization

However, for the visualization part, I'm trapped in drawing a rectangle with a relative font size by using Compose.jl. It seems that the Compose.jl does not provide a way to realize relative font size? I'm still trying to figure out this question. And I'm also taking some other graphic lib into consideration, such as [Luxor.jl](https://github.com/cormullion/Luxor.jl) /[Graphics2D.jl](https://github.com/sswatson/Graphics2D.jl)/or directly write it by [Cairo](https://github.com/JuliaLang/Cairo.jl)

Roger-luo commented:
@i2000s 

The prototype is updated to a new designed Gate type and interface, so does in proposal.

### For the proposal goal part:

I've modified part of the proposal and project goal in proposal in commit update(0ee839c) :

The Project Goal(rather than proposal goal)
I think the project's goal is to provide some layers and tool kit as you mentioned:

![structure of the project](https://github.com/Roger-luo/Resources/blob/master/GSoC/2016/imag/structure.png)
- 1st layer: basic simulators
- 2nd layer: compilers and circuit CAD tool kits
- 3rd layer: quantum programing language

my current goal will only focus on the **1st layer** and part of the **circuit CAD tool kits** and **2nd layer** as lack of knowledege and time(these contents may already fill this years GSoC) for other two layers.

This means, at least, I will finish most of the **1st layer** after GSoC2016, and the tool kits will only provide a visualization part. Constructors for circuits is updated in Kebab and I've a simple doc in both proposal and README of Kebab.

And if there is or will be a standard for APIs on quantum computer,the first layer should and will be wrapped in this standard (in the future).The 1st layer could be replaced by a real quantum computer.(this may explain the role 1st layer plays in this framework at present)

the first layer will provide a simulated environment for developers/users in the 2nd layer.

### For circuit visualization

My current plan is to realize it in Julia's graphic libraries rather in LaTeX, though LaTeX have very convenient way to draw the circuits. This is because though LaTeX is convenient on drawing circuits, but it may not be that convenient to draw other visualizations in quantum mechanics/ quantum computations (for example, Bloch Sphere, graphs, curves) as its plotting may not be as powerful as matplotlib/Gadfly/Compose/Cairo/... 

I guess there will be someone making visualization for other use in JuliaQuantum. So as the first package that offers visualization, it could be better to use a more general graphic package, which may help us reduce dependencies in the future. (It could be ugly to use both LaTeX and other graphic libraries)

And what was in my mind is to directly use the info stored by `Circuit` type to construct a picture.

### For `Circuit` type

I changed part of the design in the prototype, Kebab. And I currently think store the gates by an array of tuples could be a better way, which offers all the infos and have the simplest structure?

About the whole type system, I may need to try some types for one-way models in the prototype first, and will also take suggestion for simplification on type systems then I may find a quite appropriate type design.

And thanks for the suggestion, I will make a list of gates and their algorithms in to-do.

I've modified part of the proposal as you suggested, for other suggestions you mentioned above, I need a study of mentioned materials above and will reply them later. 

And I will ask Prof. Han about the talk. He said he read all the issues but he is not used to use issues for discussion and will help me improve my proposal and find out a time line. :-)

@amitjamadagni 

I drew the structure of this project (not the proposal) in proposal. This may make it clearer to understand the difference and similarity with other packages and software. I do not have plans on topological models in this GSoC2016, but this will surely be done after GSoC2016.(The time may not be enough, if I make a topological model's simulator as well). I think I will focus on implement these three models mentioned above and make them more efficient, and visualize commonly used part (circuit and graph state)

For the error correction and noise, I need to have a more detailed discussion with Prof. Han, will reply this soon.

About the parameters, users can tweak: this is more like I simulated a quantum computer in the first layer. So the parameters users can tweak mainly is the input states and structure of the gates. (if some specific gates have their own parameter needs to be tweaked, there will be a constructor for this gate)

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
```diff
@@ -0,0 +1,414 @@
+    + [one-way quantum computation](http://arxiv.org/abs/quant-ph/0602096v1)
+    + quantum error corrections by quantum circuits and one-way model
+- optimizations:
+    + **quantum circuit optimization** auto-optimize the circuit to an easy-realized form for experiments,this will use a program written by a former student of Prof.Han and reconstruct it in Julia
```
A reference here citing the program will be helpful !

Roger-luo replied:
added

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
```diff
@@ -0,0 +1,414 @@
+
+### Prototype
+I wrote a prototype for my idea with only simple features about quantum circuit and adiabatic quantum simulation module in [Kabab](https://github.com/Roger-luo/Kebab).
+
```
Kebab in the link.

Roger-luo replied:
@amitjamadagni Should I delete this link?

amitjamadagni replied:
Just a spell check !

Roger-luo replied:
corrected

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
````diff
@@ -0,0 +1,414 @@
+```julia
+abstract AbstractGates
+abstract AbstractCircuit
+
````
Also I got confused between `AbstractGates` and `QuantumGates`, could we change `AbstractGates` to `QCircuit` ?! 

Why not a separate type for `ComputationalModel` ?! I guess as per my understanding we have three layers 
`QuantumGates` allowing the user to define custom gates and all the standard gates are subtyped.
`QuantumCircuit` allowing the user to define the circuit
`ComputationalModel` where the user is allowed to play with the parameters of the model.  

Please do let me know if I am missing something here !

Roger-luo replied:
@amitjamadagni `AbstractGates` include computational models cause other computational models is considered to be a kind of quantum gate that input a state and output a state, which acts like a gate in the program's structure. So I called it `AbstractGates`, cause it will include all the gate-like functions/matrixes. And maybe I should change the `Gate` type to `QuGate`?

And I changed `AbstractCircuit` to `QuCircuit`

amitjamadagni replied:
I guess then we could decouple everything in the following way : 
`QuSimulation` as the supertype .... 
`AbstractGates` (this allows the user to create new gates)
        |-> `QuantumGates` (the popular gates go in here as subtypes)

`AbstractCircuit` (the circuit definition goes in here)

 `AbstractComputationalModels` (this will allow creation of user created Computational Models)
        |-> Adiabatic Model
        |-> One Way Computation

Roger-luo replied:
@amitjamadagni But I guess `AbstractComputationalModels` should be a subtype of `AbstractGates`, or we may need to define more functions for computational models, if we want to allow users combine circuits and other models. eg. combine circuit and adiabatic model. And does QuSimulation is the supertype for all the other types in this package?

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
```diff
@@ -0,0 +1,420 @@
+#### Type System
+- **Quantum Gates** `AbstractGates` is a super type for all modules in the structure of this framework. The simulations of a quantum computation model will be construct in a quantum-circuit-like way. The other quantum computation model will be a gate-like module(subtype of `AbstractGates`) in quantum circuit.
+- **Quantum Gate** quantum gates `Gate` includes gate-like modules and single qubit gates like Hadamard gate,Pauli gates...
+- **Graph** graph type will be used to store graph states for one-way model, the type is current considered to use adjacent matrix and use [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl) as dependency to visualize
```
May be we could move this as this a part of a particular model that is the one-way computation model. So I would re-structure this a bit in the following way
1. `QSimulation` -> supertype consists of `AbstractGates`, `AbstractCircuits`, `AbstractComputationalModels`
2. `AbstractGates` type might look in the following way (provide the description of different parameters) and have `QuantumGates` as subtypes (most common ones)
3. `AbstractComputationalModel` type might look in the following way (provide a general description of different parameters which might allow construction of new models). `Adiabatic`, `One-Way-Computation` are subtypes of this
4. For each of the models described that is `Adiabatic`, `One-Way-Computation` may be give an insight into what each model might deliver (like using `Graph` for `One-Way-Computation`)  

Thoughts ?!

amitjamadagni replied:
May be we could move this as this a part of a particular model that is the one-way computation model. So I would re-structure this a bit in the following way

1. `QSimulation` -> supertype consists of `AbstractGates`, `AbstractCircuits`, `AbstractComputationalModels`

2. `AbstractGates` type might look in the following way (provide the description of different parameters) and have `QuantumGates` as subtypes (most common ones)

3. `AbstractComputationalModel` type might look in the following way (provide a general description of different parameters which might allow construction of new models). `Adiabatic`, `One-Way-Computation` are subtypes of this

4. For each of the models described that is `Adiabatic`, `One-Way-Computation` may be give an insight into what each model might deliver (like using `Graph` for `One-Way-Computation`)  

Thoughts ?!

Roger-luo replied:
yup, thanks. And I think that methods in `AbstractComputationalModel` could generate a certain gate under given parameters, so that users can tweak the parameters more flexible! I'll change the structure to this.

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
```diff
@@ -0,0 +1,420 @@
+type GraphState
+    adjaMatrix::AbstractMatrix
+end
+
```
As mentioned in the above comment : I would move this to a particular model ... So the prototype can again be re-structured as above to provide more clarity ! I feel this is a bit mixed, though it seems to convey the idea. So I would say, decoupling and classifying the type definitions into three broad sections that is `AbstractGates`, `AbstractCircuit`, `AbstractComputationalModel` might be helpful.

Thoughts ?!

Roger-luo replied:
Ok, I need to attend an experiment course this afternoon, will update a new structure of prototype when I'm back in 2 hours. :-)

Roger-luo replied:
added `AbstractComputModels` and changed the role of this type in 2ef528e

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
````diff
@@ -0,0 +1,420 @@
+# For visualization, a circuit should be plotted by
+plot(QFT_circuit)
+show()
+```
````
I am able to understand these prototypes ! But could you provide a reference to `QFT` using other Computational Models like `Adiabatic Computation` or `One-Way-Computation` ?! I think that would give a complete picture of the project implementation.

Roger-luo replied:
added a reference of adiabatic computation, but maybe I need  a prototype for adiabatic computation for QFT here?

amitjamadagni replied:
A reference should do but if you can manage to work a prototype it would be nice as well :)

amitjamadagni commented on GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md:
````diff
@@ -0,0 +1,420 @@
+```
+
+# Proposal Timeline
+
````
I guess you could split the timeline into 3 parts : 
1. (Before) Community Bonding Period, 
2. Coding period (split into 2 parts : 
    i. Part - 1 : From the start to Mid-term evaluations 
   ii. Part - 2 : From Mid-term evaluations to End Term)
3. Post GSoC 

For the coding part split the major target into minor targets which could be achieved like in fewer weeks. So for example, the timeline would look like : 

Week 1 - Week 3 : 
A list of TODO tasks
Targets to be achieved : Summary of the smaller targets achieved  

Similarly for other weeks !
Splitting the timeline would give everyone much better clarity on the targets of the project. Also it would give you more clarity where you might require help and assistance !

amitjamadagni replied:
I guess we could split the timeline again on the structure of the prototype : Starting with basic functionality of `AbstractGates`, `AbstractCircuit`, `AbstractComputationalModels`. This would be the base layer for the project and then we can start constructing on this. As the main focus of the project is to support functionality for various computational models we could devote most of the time to achieve this. May be we could make this the central piece and add structures keeping in mind the broader picture, that is the long term goal.

i2000s commented:
I roughly put some text on the NumFOCUS' [GSoC idealist](https://github.com/numfocus/gsoc/blob/master/2016/ideas-list-juliaquantum.md). Both @Yongjianhan and @acroy are on the mentors' list. Mentors should register to Google before the proposal application ends. @Roger-luo if you find anything to be changed, please go ahead and commit. I am on a series of meetings today. With [Dr. Stephen Jordan](http://www.nist.gov/itl/math/sjordan.cfm) now on SS Monte Carlo simulations, and he may have some code to share online later (@amitjamdagni) . Thanks.

Roger-luo commented:
This proposal was not selected as one of the GSoC project as the enrollment letter was not in the correct form. I didn't know this massage until I checked the GSoC website manually before the last day which was too late according to the rules. So unfortunately, this project will not be eligible.

However, the project will go on, if anyone have interests in this idea, please check [QuCmp.jl](https://github.com/JuliaQuantum/QuCmp) in JuliaQuantum's repo.

To prevent the same accident from happening again, I'll leave some advise here for other students from where google is not available (perhaps China):

### For students
- If you have trouble in connecting GSoC's official website, (eg. the browser gives you 404 error), please ask JuliaQuantum members for help by email or gitter.
- check your application status after your application regularly
- if you do not use gmail very often or you have trouble in using gmail, please set your gmail automatically forward the emails to your commonly used email
- Besides your proposal, check the enrollment letter, it has to be your university's official certificate in English, which could take some time in the preparation. 

### For mentors

Since JuliaQuantum mentors may not originally come from its members, and it's possible to invite mentors in certain field, some mentors from China could be the first time in participating. So here are some suggestions for mentors from places that may have trouble in using Google services.
- You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.

### Possible tools if you want to fix the connection

[Lantern](https://github.com/getlantern/lantern), [shadowsock](https://github.com/shadowsocks/shadowsocks), and VPN services.

If you have more questions, you could also find more help in [JuliaCN](http://julialang.cn)'s QQ groups. Feel free to join and ask questions.

i2000s commented:
@Roger-luo Sorry about the application result. To my understanding, your advice to future students only higher the bar of enrollment -- some of your points may scare future applicants away. JuliaQuantum didn't aware of the fact that Google itself has announced quitting from the Chinese market a few years ago until this project. If Google doesn't want to obey the Chinese law and decides to reenter the Chinese market, I don't see there is any hope to let Chinese people use the Google service normally. We can hardly do anything to your ultimate hope. On the other hand, what JuliaQuantum can do is to provide minimal help to facilitate this process, for example, to  transfer necessary document from the Google website to the interested applicants and negotiate with the GSoC organizers for easier process, like allowing participants to use non-Gmail address to communicate. I had personally asked you if you need me to copy the GSoC rules to you before your applications, but you thought you can view those content by yourself and follow the procedure. Maybe we need to emphasize some important deadlines to future applicants to avoid the misfortune happening again. We want to avoid any miscommunication issues in the future if that was an issue. If you can provide a list of things that future JuliaQuantum organizers can do and emphasize to guarantee a successful enrollment and a smooth process on a GSoC project, that would be great. If you think in principle and from your past experience that no one in the region where Google is not available can be approved for future GSoC projects, I would personally  discourage future students from those regions to consider GSoC projects, which may not be true. JuliaQuantum will decide the police based on your experience. 

From my point of view, to all future applicants, please guarantee one mentor you choose to be able to do the code review. JuliaQuantum organizers can try our best for a limited mentor match and encourage the community to help on side but cannot support your applications if you don't find a guaranteed code reviewer to be your mentor. The mentor you choose may not be originally a JuliaQuantum member, but should has some open source project experience. After a review of the panel, we may be able to add your mentor into the org to continue your application process. 

Months after the application, I hope this is a good time that you have cooled down and seriously look back, reflect and give valuable suggestions for future successful GSoC events. Please add executable  suggestions to the README file of the proposed merge for reference to future student applicants and J/GSoC project organizers from your point of view. Thank you, and good luck to your future endeavor!

Roger-luo commented:
Do you mean the README.md file in this repo? Will it be better to have a gh-pages for this repo? I tried to add some suggestions in this file, but it looks weird with other materials, what about create a page for GSoC like JuliaLang's? @i2000s

i2000s commented:
@Roger-luo I mean the README file of your proposal folder. If you can help or write a CI script to maintain the gh-pages, feel free to create a page. Let you and me chat a little bit before any merge. Check your gitter message.

Roger-luo commented:
I added a README file first, check if there are any other problems. @i2000s  I'll squash commits later if it is OK to merge.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+# Note for 2016 GSoC application
+
+It's my first time in the GSoC application. Due to my lack of the process, I missed the deadline of updating my enrollment letter, which finally caused the eligibility.
```
It's my first time applying for the GSoC program. For lack of experiences, I missed the deadline of submitting my enrollment letter, which finally caused the disapproval.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+# Note for 2016 GSoC application
```
Notes

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+It's my first time in the GSoC application. Due to my lack of the process, I missed the deadline of updating my enrollment letter, which finally caused the eligibility.
+
+This proposal was not selected as one of the GSoC project as the enrollment letter was not in the correct form. I didn't know this massage until I checked the GSoC website manually before the last day which was too late according to the rules. So unfortunately, this project will not be eligible.
```
~~This proposal was not selected as one of the GSoC project as the enrollment letter was not in the correct form.~~ I didn't know the requirement of enrollment letter until I checked the GSoC website one day before the deadline. Unfortunately, this proposal was not approved only because of the miss of deadline in submitting a supporting letter. But the proposal itself has passed organizational peer reviews and could be useful for future reference.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+This proposal was not selected as one of the GSoC project as the enrollment letter was not in the correct form. I didn't know this massage until I checked the GSoC website manually before the last day which was too late according to the rules. So unfortunately, this project will not be eligible.
+
+However, the project will go on, if anyone have interests in this idea, please check [QuCmp.jl](https://github.com/JuliaQuantum/QuCmp) in JuliaQuantum's repo.
```
In any case, the project has to move on. For your interest, please check [QuCmp.jl](https://github.com/JuliaQuantum/QuCmp) in JuliaQuantum's repo.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+However, the project will go on, if anyone have interests in this idea, please check [QuCmp.jl](https://github.com/JuliaQuantum/QuCmp) in JuliaQuantum's repo.
+
+To prevent the same accident from happening again and to help people in region without Google services to get involved with all those amazing ideas, I'll leave some advise here for other students from where google is not available :
```
To prevent the same tragedy happening on future applicants, especially to help applicants in regions without Google services, I am leaving some advice below.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+### For students
+
+- If you have trouble in connecting GSoC's official website, (eg. the browser gives you 404 error), please ask JuliaQuantum members for help by email or gitter.
```
- If you have trouble in connecting GSoC's official website, eg. the browser gives you 404 error, you should ask JuliaQuantum members for help by email or gitter. Make sure you have read the rules and regulations before applying, with all the deadlines in mind.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+### For students
+
+- If you have trouble in connecting GSoC's official website, (eg. the browser gives you 404 error), please ask JuliaQuantum members for help by email or gitter.
+- check your application status after your application regularly
```
- Check your application status regularly after your application.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+- If you have trouble in connecting GSoC's official website, (eg. the browser gives you 404 error), please ask JuliaQuantum members for help by email or gitter.
+- check your application status after your application regularly
+- if you do not use gmail very often or you have trouble in using gmail, please set your gmail automatically forward the emails to your commonly used email
```
- If you do not use gmail very often or you can hardly use gmail, please set your gmail to automatically forward emails to your commonly used email address. The organizers from Google will send emails to your Gmail address by default.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+- If you have trouble in connecting GSoC's official website, (eg. the browser gives you 404 error), please ask JuliaQuantum members for help by email or gitter.
+- check your application status after your application regularly
+- if you do not use gmail very often or you have trouble in using gmail, please set your gmail automatically forward the emails to your commonly used email
+- Besides your proposal, check the enrollment letter, it has to be your university's official certificate in English, which could take some time in the preparation.
```
- Prepare your enrollment letter early before the deadline, which could take some time and has to be your university's official certificate in English.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+### For mentors
+
+Since JuliaQuantum mentors may not originally come from its present members, and it's possible to invite people as your mentor, some mentors from China could be the first time in participating. So here are some suggestions for mentors from places that may have trouble in using Google services.
```
By courtesy of JuliaQuantum's policy that non-organizational person can join the organization by passing a qualification review process and act as a mentor of the applicant, it is possible for outsiders to become a GSoC mentor under the organization's umbrella. Again, if there is any communication problems with Google or understanding the rules, the organization could offer a hand.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+Since JuliaQuantum mentors may not originally come from its present members, and it's possible to invite people as your mentor, some mentors from China could be the first time in participating. So here are some suggestions for mentors from places that may have trouble in using Google services.
+
+- You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.
```
Delete.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+- You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.
+
+
+### Possible tools if you want to fix the connection
```
### Useful tools for fixing connection problems with Google servers on your own

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+### Possible tools if you want to fix the connection
+
+[Lantern](https://github.com/getlantern/lantern), [shadowsock](https://github.com/ziggear/shadowsocks), and VPN services.
```
[Lantern](https://github.com/getlantern/lantern), [shadowsock](https://github.com/ziggear/shadowsocks), and VPN services. Use at your own risk.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,29 @@
+
+[Lantern](https://github.com/getlantern/lantern), [shadowsock](https://github.com/ziggear/shadowsocks), and VPN services.
+
+If you have more questions, you could also find more help in [JuliaCN](http://julialang.cn)'s QQ groups. Feel free to join and ask questions.
```
For people living in China, you may also find [JuliaCN](http://julialang.cn)'s QQ groups helpful. Feel free to join and ask questions there.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -1,32 +1,33 @@
-Since JuliaQuantum mentors may not originally come from its present members, and it's possible to invite people as your mentor, some mentors from China could be the first time in participating. So here are some suggestions for mentors from places that may have trouble in using Google services.
+By courtesy of JuliaQuantum's policy that non-organizational person can join the organization by passing a qualification review process and act as a mentor of the applicant, it is possible for outsiders to become a GSoC mentor under the organization's umbrella. Again, if there is any communication problems with Google or understanding the rules, the organization could offer a hand.
 
 - You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.
```
~~\- You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.~~

i2000s commented:
Could you rename the file `GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md` without the ":" sign? That special character is illegal to use in file names in Windows. Thanks. 

After all these changes, I would merge it directly to the repo without a squash to keep the history of edit intact.

i2000s commented on GSoC/2016/README.md:
```diff
@@ -0,0 +1,33 @@
+
+By courtesy of JuliaQuantum's policy that non-organizational person can join the organization by passing a qualification review process and act as a mentor of the applicant, it is possible for outsiders to become a GSoC mentor under the organization's umbrella. Again, if there is any communication problems with Google or understanding the rules, the organization could offer a hand.
+
+- You may have trouble in reading through the mentor's guide and rules in GSoC's official website. If so, please feel free to ask in JuliaQuantum's email list or gitter channel for help.
```
Don't forget to delete this paragraph as I have commented earlier. I think it's repeating and over-shooting. Thanks.
